crates/metadata/src/compatibility.rs

```diff
@@ -0,0 +1,177 @@
+// Copyright 2018-2023 Parity Technologies (UK) Ltd.
+// This file is part of cargo-contract.
+//
+// cargo-contract is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// cargo-contract is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with cargo-contract.  If not, see <http://www.gnu.org/licenses/>.
+
+use anyhow::{
+    anyhow,
+    bail,
+    Result,
+};
+use semver::{
+    BuildMetadata,
+    Comparator,
+    Op,
+    Version,
+    VersionReq,
+};
+
+/// Version of the currently executing `cargo-contract` binary.
+const VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[derive(Debug, serde::Deserialize)]
+struct CompatibilityList {
+    versions: Vec<CompatibilityInfo>,
+}
+
+#[derive(Debug, serde::Deserialize)]
+struct CompatibilityInfo {
+    #[serde(rename = "cargo-contract")]
+    cargo_contract_req: VersionReq,
+    #[serde(rename = "ink")]
+    ink_req: VersionReq,
+}
+
+trait CustomMatch {
+    fn matches_including_pre_releases(&self, version: &Version) -> bool;
+}
+
+// The VersionReq matching mechanism behaves in such a way that 4.0.0-alpha.1 does not
+// satisfy the requirement ">2.0.0,<5.0.0". To modify this behavior, a custom matching
+// implementation has been provided.
```

Comment:
As discussed, I wonder if it is better just to deal in non pre-release version numbers. So for the configured specs we don't allow pre-release versions, and for the versions of `cargo-contract` and `ink` we just set the `pre` to `None`. WDYT?

Reply:
The issue with this solution is that information loss can occur if compatibility changes have been introduced between certain pre-release versions, as seen in the example of Ink 4.0.0-alpha3. Therefore, if someone utilizes Ink 4.0.0-alpha1 in their contract, we may mistakenly assume it to be version 4.0.0 and therefore compatible with cargo-contract 2.0.0, when in fact, it is not.

Reply:
If we apply the two other suggestions about the spec, we will no longer require the custom matching and can stay true to the semver ranges, which I think is important to avoid confusion.

Reply:
I think it should work, but the downside is that it will result in higher maintenance costs because for each release, you'll need to add a new entry. Previously, you could skip it until a breaking change was introduced.